### PR TITLE
[IMP] stock: reserved quantity put in pack

### PR DESCRIPTION
**Pull request by jip-odoo**
Task: https://www.odoo.com/web?#id=1966999&action=327&model=project.task&view_type=form&menu_id=4720
Pad: https://pad.odoo.com/p/r.444df3af7ad97ea97d851c6cf3dbaa37

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/stock/models/stock_picking.py [hunk 1]
```diff
@@ -1026,7 +1026,7 @@ def _put_in_pack(self, move_line_ids):
             package = self.env['stock.quant.package'].create({})
             for ml in move_line_ids:
                 if float_compare(ml.qty_done, ml.product_uom_qty,
-                                 precision_rounding=ml.product_uom_id.rounding) >= 0:
+                                 precision_rounding=ml.product_uom_id.rounding) >= 0 or ml.qty_done == 0.0:
```
Comment (sle-odoo) [inline addons/stock/models/stock_picking.py]: this is ugly as hell + you don't even use float_compare

Comment (sle-odoo) [reply addons/stock/models/stock_picking.py]: in put_in_pack, either:
- you have some quantities done anywhere and you keep the old behaviour
- you don't have any quantities done at all BUT you some moves are reserved, then set the qty done as the qty reserved on the move line and call the old code

File: addons/stock/tests/test_move.py [hunk 2]
```diff
@@ -4186,16 +4186,14 @@ def test_put_in_pack_1(self):
...
-        unpacked_ml.qty_done = 1
-        picking.put_in_pack()
-        self.assertEqual(len(picking.move_line_ids), 2)
-        self.assertEqual(self.env['stock.quant']._get_available_quantity(self.product, self.stock_location), 0)
-        picking.button_validate()
-        self.assertEqual(self.env['stock.quant']._get_available_quantity(self.product, self.stock_location), 0)
+        res_dict = picking.button_validate()
+        self.assertEqual(res_dict.get('res_model'), 'stock.immediate.transfer')
+        wizard = self.env[(res_dict.get('res_model'))].browse(res_dict.get('res_id'))
+        wizard.process()
```
Comment (sle-odoo) [inline addons/stock/tests/test_move.py]: keep this old test and add a new one, here you changed what was tested and i don't see why you would need an immediate transfer anywhere?

test1)
one move of product 1 reserved
one move of product 2 reserved 

hit put in pack
you have two move line qty done 1 result package package1

test2)
one move of product 1 reserved
one move of product 2 reserved 
product 1 set qty done 1
put in pack
product 2 set qty done 2
put in pack
you have two move line qty done 1 result package package1 and package2

that would be a start

File: addons/stock/models/stock_picking.py (hunk 1 above)
Comment (jip-odoo) [reply addons/stock/models/stock_picking.py]: Done

File: addons/stock/tests/test_move.py (hunk 2 above)
Comment (jip-odoo) [reply addons/stock/tests/test_move.py]: Done

File: addons/stock/models/stock_move.py [hunk 3]
```diff
@@ -1040,6 +1040,8 @@ def _update_reserved_quantity(self, need, available_quantity, location_id, lot_i
             to_update = self.move_line_ids.filtered(lambda ml: ml._reservation_is_updatable(quantity, reserved_quant))
             if to_update:
                 to_update[0].with_context(bypass_reservation_update=True).product_uom_qty += self.product_id.uom_id._compute_quantity(quantity, to_update[0].product_uom_id, rounding_method='HALF-UP')
+                if to_update[0].result_package_id:
+                    to_update[0].qty_done = to_update[0].product_uom_qty
```
Comment (sle-odoo) [inline addons/stock/models/stock_move.py]: what? why?

Comment (jip-odoo) [reply addons/stock/models/stock_move.py]: I consider this case, If 80 as the initial quantity and 60 as a reserved quantity and if we put in the pack then done quantity is 60 after that when 20 quantity is available and then we validate the transfer then backorder is created.
but this solution is not valid so, I will revert this.

**Comment by sle-odoo** [comment]
robodoo r+

**Comment by julianstefan** [comment]
this change should be a new setting that can be enabled, instead of changing the default bevavior. it can be undesired for certain use cases.
Imagine a picking in a big company with more than 50 lines, and someone uses the put in pack action by accident. all of the open lines will be packed and all progress will be lost.

**Pull request closed without merging**